File: proxy-android/proxy/src/main/java/com/zeticai/melangelm/MelangeLmProxy.kt
package com.zeticai.melangelm

import android.content.Context
import com.zeticai.melangelm.model.ChatMessage
import com.zeticai.melangelm.model.ChatRequest
import com.zeticai.melangelm.model.PipelineOnlyResult
import com.zeticai.melangelm.model.ProxyResult
import com.zeticai.melangelm.pipeline.PipelineStage
import com.zeticai.melangelm.pipeline.ProxyPipeline
import com.zeticai.melangelm.stages.AnonymizerStage
import com.zeticai.melangelm.stages.PromptGuardStage
import com.zeticai.melangelm.stages.SummarizerStage
import com.zeticai.melangelm.upstream.OpenAIUpstreamClient
import com.zeticai.melangelm.upstream.UpstreamClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * # Melange LM Proxy
 *
 * An on-device middleware layer that sits between your app and any OpenAI-compatible
 * LLM API. It runs a configurable pipeline of on-device models before every request:
 *
 * ```
 * User message
 *     │
 *     ▼
 * ┌─────────────────────────────────────┐
 * │  PromptGuard  (block malicious)     │  on-device
 * │  TextAnonymizer (redact PII)        │  on-device
 * │  Summarizer (reduce tokens) [soon]  │  on-device
 * └─────────────────────────────────────┘
 *     │
 *     ▼  (only clean, anonymized, compressed prompt reaches the API)
 * Upstream LLM API (OpenAI / Azure / etc.)
 *     │
 *     ▼
 * ┌─────────────────────────────────────┐
 * │  De-anonymize response              │  on-device
 * └─────────────────────────────────────┘
 *     │
 *     ▼
 * Your app
 * ```
 *
 * ## Usage
 *
 * ```kotlin
 * val proxy = MelangeLmProxy.build(context) {
 *     promptGuard {
 *         personalKey = "your_zetic_key"
 *     }
 *     anonymizer {
 *         personalKey = "your_zetic_key"
 *         restoreInResponse = true
 *     }
 *     upstream {
 *         baseUrl = "https://api.openai.com"
 *         apiKey = BuildConfig.OPENAI_API_KEY
 *     }
 * }
 *
 * // Initialize (loads on-device models) — do this at app startup
 * proxy.initialize()
 *
 * // Chat
 * val result = proxy.chat(
 *     messages = listOf(ChatMessage("user", "Hello, what's my account balance?"))
 * )
 * when (result) {
 *     is ProxyResult.Success -> println(result.response.choices.first().message.content)
 *     is ProxyResult.Blocked -> println("Blocked: ${result.reason} by ${result.stage}")
 *     is ProxyResult.Error   -> println("Error: ${result.message}")
 * }
 * ```
 */
class MelangeLmProxy private constructor(
    private val pipeline: ProxyPipeline,
    private val defaultModel: String
) {
    private val scope = CoroutineScope(Dispatchers.IO)

    /**
     * Load all on-device models. Call this once at app startup (e.g. in Application.onCreate).
     * Subsequent calls are no-ops.
     * @param onStageReady Called with each stage's name as it finishes loading.
     * @param onStageProgress Called with (stageName, progress 0.0–1.0) during model download.
     */
    suspend fun initialize(
        onStageReady: ((String) -> Unit)? = null,
        onStageProgress: ((String, Float) -> Unit)? = null
    ) = pipeline.initialize(onStageReady, onStageProgress)

    /** Update the summarizer's compression target ratio without rebuilding the proxy. */
    fun updateCompressionRatio(ratio: Double) = pipeline.updateCompressionRatio(ratio)

    /**
     * Run only the on-device pipeline stages (no upstream LLM call).
     * Useful for demo mode or previewing what the pipeline does to messages.
     */
    suspend fun processOnly(
        messages: List<ChatMessage>,
        model: String = defaultModel
    ): PipelineOnlyResult = pipeline.processOnly(
        ChatRequest(model = model, messages = messages)
    )

    /**
     * Run a chat request through the full pipeline and return the result.
     */
    suspend fun chat(
        messages: List<ChatMessage>,
        model: String = defaultModel,
        temperature: Double? = null,
        maxTokens: Int? = null
    ): ProxyResult = pipeline.process(
        ChatRequest(
            model = model,
            messages = messages,
            temperature = temperature,
            maxTokens = maxTokens
        )
    )

    // -------------------------------------------------------------------------
    // Builder DSL
    // -------------------------------------------------------------------------

    class Builder(private val context: Context) {
        private var promptGuardConfig: PromptGuardConfig? = null
        private var anonymizerConfig: AnonymizerConfig? = null
        private var summarizerConfig: SummarizerConfig? = null
        private var upstreamConfig: UpstreamConfig = UpstreamConfig()
        private val extraStages = mutableListOf<PipelineStage>()

        fun promptGuard(block: PromptGuardConfig.() -> Unit) {
            promptGuardConfig = PromptGuardConfig().apply(block)
        }

        fun anonymizer(block: AnonymizerConfig.() -> Unit) {
            anonymizerConfig = AnonymizerConfig().apply(block)
        }

        fun summarizer(
            personalKey: String,
            modelId: String = "yeonseok_zeticai_ceo/LFM2-comparison",
            llmTarget: com.zeticai.melangelm.stages.LLMTarget = com.zeticai.melangelm.stages.LLMTarget.LLAMA_CPP,
            llmQuantType: com.zeticai.melangelm.stages.LLMQuantType = com.zeticai.melangelm.stages.LLMQuantType.Q4,
            minCharsToSummarize: Int = 300,
            compressionTargetRatio: Double = 0.5
        ) {
            summarizerConfig = SummarizerConfig(
                personalKey = personalKey,
                modelId = modelId,
                llmTarget = llmTarget,
                llmQuantType = llmQuantType,
                minCharsToSummarize = minCharsToSummarize,
                compressionTargetRatio = compressionTargetRatio
            )
        }

        fun upstream(block: UpstreamConfig.() -> Unit) {
            upstreamConfig = UpstreamConfig().apply(block)
        }

        /** Add a custom pipeline stage. Stages run in insertion order after built-in stages. */
        fun addStage(stage: PipelineStage) {
            extraStages.add(stage)
        }

        fun build(): MelangeLmProxy {
            val stages = mutableListOf<PipelineStage>()

            promptGuardConfig?.let { cfg ->
                stages.add(
                    PromptGuardStage(
                        context = context,
                        personalKey = cfg.personalKey,
                        maliciousThreshold = cfg.maliciousThreshold,
                        checkRoles = cfg.checkRoles
                    )
                )
            }

            anonymizerConfig?.let { cfg ->
                stages.add(
                    AnonymizerStage(
                        context = context,
                        personalKey = cfg.personalKey,
                        redactRoles = cfg.redactRoles,
                        restoreInResponse = cfg.restoreInResponse
                    )
                )
            }

            summarizerConfig?.let { cfg ->
                stages.add(
                    SummarizerStage(
                        context = context,
                        personalKey = cfg.personalKey,
                        modelId = cfg.modelId,
                        llmTarget = cfg.llmTarget,
                        llmQuantType = cfg.llmQuantType,
                        minCharsToSummarize = cfg.minCharsToSummarize,
                        compressionTargetRatio = cfg.compressionTargetRatio
                    )
                )
            }

            stages.addAll(extraStages)

            val upstream: UpstreamClient = upstreamConfig.customClient
                ?: OpenAIUpstreamClient(
                    baseUrl = upstreamConfig.baseUrl,
                    apiKey = upstreamConfig.apiKey,
                    defaultModel = upstreamConfig.defaultModel,
                    timeoutSeconds = upstreamConfig.timeoutSeconds
                )

            return MelangeLmProxy(
                pipeline = ProxyPipeline(stages, upstream),
                defaultModel = upstreamConfig.defaultModel ?: "gpt-4o-mini"
            )
        }
    }

    data class PromptGuardConfig(
        var personalKey: String = "",
        var maliciousThreshold: Float = 0f,
        var checkRoles: Set<String> = setOf("user")
    )

    data class AnonymizerConfig(
        var personalKey: String = "",
        var redactRoles: Set<String> = setOf("user"),
        var restoreInResponse: Boolean = true
    )

    data class SummarizerConfig(
        var personalKey: String = "",
        var modelId: String = "yeonseok_zeticai_ceo/LFM2-comparison",
        var llmTarget: com.zeticai.melangelm.stages.LLMTarget = com.zeticai.melangelm.stages.LLMTarget.LLAMA_CPP,
        var llmQuantType: com.zeticai.melangelm.stages.LLMQuantType = com.zeticai.melangelm.stages.LLMQuantType.Q4,
        var minCharsToSummarize: Int = 300,
        var compressionTargetRatio: Double = 0.5
    )

    data class UpstreamConfig(
        var baseUrl: String = "https://api.openai.com",
        var apiKey: String = "",
        var defaultModel: String? = null,
        var timeoutSeconds: Long = 60,
        var customClient: UpstreamClient? = null
    )

    companion object {
        /**
         * Build a proxy with full control over every stage.
         */
        fun build(context: Context, block: Builder.() -> Unit): MelangeLmProxy =
            Builder(context).apply(block).build()

        /**
         * Quick setup: all three on-device stages (PromptGuard + Anonymizer + Summarizer).
         * Best balance of safety, privacy, and cost savings.
         */
        fun allFeatures(
            context: Context,
            zeticKey: String,
            baseUrl: String = "https://api.openai.com",
            apiKey: String,
            model: String = "gpt-4o-mini",
            compressionTarget: Double = 0.5
        ): MelangeLmProxy = build(context) {
            promptGuard { personalKey = zeticKey }
            anonymizer { personalKey = zeticKey; restoreInResponse = true }
            summarizer(personalKey = zeticKey, compressionTargetRatio = compressionTarget)
            upstream { this.baseUrl = baseUrl; this.apiKey = apiKey; defaultModel = model }
        }

        /**
         * Safety-only setup: PromptGuard + Anonymizer, no summarization.
         * No token compression — use when response quality matters more than cost.
         */
        fun safetyOnly(
            context: Context,
            zeticKey: String,
            baseUrl: String = "https://api.openai.com",
            apiKey: String,
            model: String = "gpt-4o-mini"
        ): MelangeLmProxy = build(context) {
            promptGuard { personalKey = zeticKey }
            anonymizer { personalKey = zeticKey; restoreInResponse = true }
            upstream { this.baseUrl = baseUrl; this.apiKey = apiKey; defaultModel = model }
        }

        /**
         * Cost-optimized setup: Summarizer only, no safety/privacy stages.
         * Maximum token savings with no PII redaction overhead.
         */
        fun costOptimized(
            context: Context,
            zeticKey: String,
            baseUrl: String = "https://api.openai.com",
            apiKey: String,
            model: String = "gpt-4o-mini",
            compressionTarget: Double = 0.5
        ): MelangeLmProxy = build(context) {
            summarizer(personalKey = zeticKey, compressionTargetRatio = compressionTarget)
            upstream { this.baseUrl = baseUrl; this.apiKey = apiKey; defaultModel = model }
        }

        /**
         * Wrap an existing UpstreamClient with on-device stages.
         * Use this when you already have your own HTTP client / Retrofit service
         * and just want to add Melange's pipeline in front of it.
         */
        fun wrap(
            context: Context,
            zeticKey: String,
            client: UpstreamClient,
            block: Builder.() -> Unit = {
                promptGuard { personalKey = zeticKey }
                anonymizer { personalKey = zeticKey; restoreInResponse = true }
                summarizer(personalKey = zeticKey)
            }
        ): MelangeLmProxy = build(context) {
            block()
            upstream { customClient = client }
        }
    }
}

File: proxy-ios/Sources/MelangeLmProxy/MelangeLmProxy.swift
//
//  MelangeLmProxy.swift
//  MelangeLmProxy
//
//  # Melange LM Proxy (iOS)
//
//  On-device middleware for LLM API calls — safer, cheaper, and privacy-preserving.
//
//  Pipeline:
//    User message → PromptGuard → TextAnonymizer → Summarizer → Upstream LLM
//                                                             ← De-anonymize ←
//

import Foundation

/// # Melange LM Proxy
///
/// An on-device middleware layer between your iOS app and any OpenAI-compatible LLM API.
///
/// ## Usage
///
/// ```swift
/// let proxy = MelangeLmProxy.build {
///     $0.promptGuard(personalKey: "your_zetic_key")
///     $0.anonymizer(personalKey: "your_zetic_key", restoreInResponse: true)
///     $0.summarizer(personalKey: "your_zetic_key")
///     $0.upstream(baseURL: "https://api.openai.com", apiKey: "sk-...")
/// }
///
/// // Initialize once at app startup (loads on-device models)
/// try await proxy.initialize()
///
/// // Use like a normal chat client
/// let result = await proxy.chat(messages: [
///     ChatMessage(role: "system", content: "You are a helpful assistant."),
///     ChatMessage(role: "user", content: "My name is Alice. What is 2+2?")
/// ])
///
/// switch result {
/// case .success(let response):
///     print(response.choices.first?.message.content ?? "")
/// case .blocked(let reason, let stage):
///     print("Blocked by \(stage): \(reason)")
/// case .failure(let message, _):
///     print("Error: \(message)")
/// }
/// ```
public final class MelangeLmProxy: @unchecked Sendable {
    private let pipeline: ProxyPipeline
    private let defaultModel: String

    private init(pipeline: ProxyPipeline, defaultModel: String) {
        self.pipeline = pipeline
        self.defaultModel = defaultModel
    }

    // MARK: - Public API

    /// Load all on-device models. Call once at app startup.
    /// Safe to call multiple times — subsequent calls are no-ops.
    /// - Parameter onStageReady: Called with each stage's name as it finishes loading.
    /// - Parameter onStageProgress: Called with (stageName, progress 0.0–1.0) during model download.
    public func initialize(
        onStageReady: ((String) -> Void)? = nil,
        onStageProgress: ((String, Float) -> Void)? = nil
    ) async throws {
        await pipeline.initialize(onStageReady: onStageReady, onStageProgress: onStageProgress)
    }

    /// Update the summarizer's compression target ratio without rebuilding the proxy.
    public func updateCompressionRatio(_ ratio: Double) {
        pipeline.updateCompressionRatio(ratio)
    }

    /// Run only the on-device pipeline stages (PromptGuard, Anonymizer, Summarizer)
    /// without calling the upstream LLM. Useful for demo mode or previewing pipeline effects.
    public func processOnly(
        messages: [ChatMessage],
        model: String? = nil
    ) async -> PipelineOnlyResult {
        let request = ChatRequest(model: model ?? defaultModel, messages: messages)
        return await pipeline.processOnly(request)
    }

    /// Run a chat request through the full pipeline and return the result.
    public func chat(
        messages: [ChatMessage],
        model: String? = nil,
        temperature: Double? = nil,
        maxTokens: Int? = nil
    ) async -> ProxyResult {
        let request = ChatRequest(
            model: model ?? defaultModel,
            messages: messages,
            temperature: temperature,
            maxTokens: maxTokens
        )
        return await pipeline.process(request)
    }

    // MARK: - Builder

    /// Build a proxy with full control over every stage.
    public static func build(_ configure: (Builder) -> Void) -> MelangeLmProxy {
        let builder = Builder()
        configure(builder)
        return builder.build()
    }

    /// Quick setup: all three on-device stages (PromptGuard + Anonymizer + Summarizer).
    public static func allFeatures(
        zeticKey: String,
        baseURL: String = "https://api.openai.com",
        apiKey: String,
        model: String = "gpt-4o-mini",
        compressionTarget: Double = 0.5
    ) -> MelangeLmProxy {
        build {
            $0.promptGuard(personalKey: zeticKey)
            $0.anonymizer(personalKey: zeticKey, restoreInResponse: true)
            $0.summarizer(personalKey: zeticKey, compressionTargetRatio: compressionTarget)
            $0.upstream(baseURL: baseURL, apiKey: apiKey, defaultModel: model)
        }
    }

    /// Safety-only setup: PromptGuard + Anonymizer, no summarization.
    public static func safetyOnly(
        zeticKey: String,
        baseURL: String = "https://api.openai.com",
        apiKey: String,
        model: String = "gpt-4o-mini"
    ) -> MelangeLmProxy {
        build {
            $0.promptGuard(personalKey: zeticKey)
            $0.anonymizer(personalKey: zeticKey, restoreInResponse: true)
            $0.upstream(baseURL: baseURL, apiKey: apiKey, defaultModel: model)
        }
    }

    /// Cost-optimized setup: Summarizer only, no safety/privacy stages.
    public static func costOptimized(
        zeticKey: String,
        baseURL: String = "https://api.openai.com",
        apiKey: String,
        model: String = "gpt-4o-mini",
        compressionTarget: Double = 0.5
    ) -> MelangeLmProxy {
        build {
            $0.summarizer(personalKey: zeticKey, compressionTargetRatio: compressionTarget)
            $0.upstream(baseURL: baseURL, apiKey: apiKey, defaultModel: model)
        }
    }

    /// Wrap an existing UpstreamClient with on-device stages.
    /// Use when you already have your own HTTP client and want to add Melange's pipeline.
    public static func wrap(
        zeticKey: String,
        client: any UpstreamClient,
        configure: ((Builder) -> Void)? = nil
    ) -> MelangeLmProxy {
        build { builder in
            if let configure = configure {
                configure(builder)
            } else {
                builder.promptGuard(personalKey: zeticKey)
                builder.anonymizer(personalKey: zeticKey, restoreInResponse: true)
                builder.summarizer(personalKey: zeticKey)
            }
            builder.setCustomClient(client)
        }
    }

    public final class Builder {
        private var promptGuardConfig: PromptGuardConfig?
        private var anonymizerConfig: AnonymizerConfig?
        private var summarizerConfig: SummarizerConfig?
        private var upstreamConfig = UpstreamConfig()
        private var extraStages: [any PipelineStage] = []

        public func promptGuard(
            personalKey: String,
            maliciousThreshold: Float = 0,
            checkRoles: Set<String> = ["user"]
        ) {
            promptGuardConfig = PromptGuardConfig(
                personalKey: personalKey,
                maliciousThreshold: maliciousThreshold,
                checkRoles: checkRoles
            )
        }

        public func anonymizer(
            personalKey: String,
            redactRoles: Set<String> = ["user"],
            restoreInResponse: Bool = true
        ) {
            anonymizerConfig = AnonymizerConfig(
                personalKey: personalKey,
                redactRoles: redactRoles,
                restoreInResponse: restoreInResponse
            )
        }

        public func summarizer(
            personalKey: String,
            modelId: String = "yeonseok_zeticai_ceo/LFM2-comparison",
            llmTarget: ProxyLLMTarget = .llamaCpp,
            llmQuantType: ProxyLLMQuantType = .q4,
            minCharsToSummarize: Int = 300,
            compressionTargetRatio: Double = 0.5
        ) {
            summarizerConfig = SummarizerConfig(
                personalKey: personalKey,
                modelId: modelId,
                llmTarget: llmTarget,
                llmQuantType: llmQuantType,
                minCharsToSummarize: minCharsToSummarize,
                compressionTargetRatio: compressionTargetRatio
            )
        }

        public func upstream(
            baseURL: String,
            apiKey: String,
            defaultModel: String = "gpt-4o-mini",
            timeoutSeconds: Double = 60
        ) {
            upstreamConfig = UpstreamConfig(
                baseURL: baseURL,
                apiKey: apiKey,
                defaultModel: defaultModel,
                timeoutSeconds: timeoutSeconds
            )
        }

        /// Add a fully custom pipeline stage after the built-in ones.
        public func addStage(_ stage: any PipelineStage) {
            extraStages.append(stage)
        }

        /// Use a custom upstream client instead of the built-in OpenAI client.
        public func setCustomClient(_ client: any UpstreamClient) {
            upstreamConfig.customClient = client
        }

        fileprivate func build() -> MelangeLmProxy {
            var stages: [any PipelineStage] = []

            if let cfg = promptGuardConfig {
                stages.append(PromptGuardStage(
                    personalKey: cfg.personalKey,
                    maliciousThreshold: cfg.maliciousThreshold,
                    checkRoles: cfg.checkRoles
                ))
            }

            if let cfg = anonymizerConfig {
                stages.append(AnonymizerStage(
                    personalKey: cfg.personalKey,
                    redactRoles: cfg.redactRoles,
                    restoreInResponse: cfg.restoreInResponse
                ))
            }

            if let cfg = summarizerConfig {
                stages.append(SummarizerStage(
                    personalKey: cfg.personalKey,
                    modelId: cfg.modelId,
                    llmTarget: cfg.llmTarget,
                    llmQuantType: cfg.llmQuantType,
                    minCharsToSummarize: cfg.minCharsToSummarize,
                    compressionTargetRatio: cfg.compressionTargetRatio
                ))
            }

            stages.append(contentsOf: extraStages)

            let upstream = upstreamConfig.customClient ?? OpenAIUpstreamClient(
                baseURL: upstreamConfig.baseURL,
                apiKey: upstreamConfig.apiKey,
                defaultModel: upstreamConfig.defaultModel,
                timeoutSeconds: upstreamConfig.timeoutSeconds
            )

            return MelangeLmProxy(
                pipeline: ProxyPipeline(stages: stages, upstream: upstream),
                defaultModel: upstreamConfig.defaultModel ?? "gpt-4o-mini"
            )
        }
    }

    // MARK: - Config structs

    struct PromptGuardConfig {
        var personalKey: String
        var maliciousThreshold: Float
        var checkRoles: Set<String>
    }

    struct AnonymizerConfig {
        var personalKey: String
        var redactRoles: Set<String>
        var restoreInResponse: Bool
    }

    struct SummarizerConfig {
        var personalKey: String
        var modelId: String
        var llmTarget: ProxyLLMTarget
        var llmQuantType: ProxyLLMQuantType
        var minCharsToSummarize: Int
        var compressionTargetRatio: Double
    }

    struct UpstreamConfig {
        var baseURL: String = "https://api.openai.com"
        var apiKey: String = ""
        var defaultModel: String? = nil
        var timeoutSeconds: Double = 60
        var customClient: (any UpstreamClient)? = nil
    }
}
